Changes in module 'nf-core/antismash/antismashlitedownloaddatabases'
--- modules/nf-core/antismash/antismashlitedownloaddatabases/main.nf
+++ modules/nf-core/antismash/antismashlitedownloaddatabases/main.nf
@@ -35,7 +35,7 @@
 
     script:
     def args = task.ext.args ?: ''
-    cp_cmd = session.config.conda.enabled ? "cp -r \$(python -c 'import antismash;print(antismash.__file__.split(\"/__\")[0])') antismash_dir;" : "cp -r /usr/local/lib/python3.8/site-packages/antismash antismash_dir;"
+    cp_cmd = workflow.containerEngine == null ? "cp -r \$(python -c 'import antismash;print(antismash.__file__.split(\"/__\")[0])') antismash_dir;" : "cp -r /usr/local/lib/python3.8/site-packages/antismash antismash_dir;"
     """
     download-antismash-databases \\
         --database-dir antismash_db \\
@@ -51,7 +51,7 @@
 
     stub:
     def args = task.ext.args ?: ''
-    cp_cmd = session.config.conda.enabled ? "cp -r \$(python -c 'import antismash;print(antismash.__file__.split(\"/__\")[0])') antismash_dir;" : "cp -r /usr/local/lib/python3.8/site-packages/antismash antismash_dir;"
+    cp_cmd = workflow.containerEngine == null ? "cp -r \$(python -c 'import antismash;print(antismash.__file__.split(\"/__\")[0])') antismash_dir;" : "cp -r /usr/local/lib/python3.8/site-packages/antismash antismash_dir;"
     """
     echo "download-antismash-databases --database-dir antismash_db $args"
 

************************************************************
